File: DynamicPrograming/js/canConcat.js
const canConcat = (s, words, memo={}) => {
    if(s in memo) return memo[s]
    if (s === "") return true

    for(let word of words){
        if(s.startsWith(word) === true){
            const suffix = s.slice(word.length)

            if(canConcat(suffix, words, memo) === true){
                memo[s] = true
                return true
            }
        }
    }

    memo[s] = false
    return false
}

const canConcat = (s, words, i = 0, memo = {}) => {
  if (i in memo) return memo[i];
  
  if (i === s.length) return true;
  
  for (let word of words) {
    if (s.startsWith(word, i)) {
      if (canConcat(s, words, i + word.length, memo)) {
        memo[i] = true;
        return true
      }
    }
  }
  
  memo[i] = false;
  return false;
};

File: DynamicPrograming/csharp/CanConcat.cs
using System.Collections.Generic;

namespace DynamicProgrammingSolutions;

public static class CanConcat
{
    public static bool CanConcatenate(string s, IReadOnlyList<string> words)
    {
        var memo = new Dictionary<int, bool>();
        return Dfs(0, s, words, memo);
    }

    private static bool Dfs(int index, string s, IReadOnlyList<string> words, IDictionary<int, bool> memo)
    {
        if (index == s.Length)
        {
            return true;
        }

        if (memo.TryGetValue(index, out var cached))
        {
            return cached;
        }

        foreach (var word in words)
        {
            if (s.AsSpan(index).StartsWith(word) && Dfs(index + word.Length, s, words, memo))
            {
                return memo[index] = true;
            }
        }

        return memo[index] = false;
    }
}
